### PR TITLE
Extend class loader table to support JITServer AOT cache

### PR DESCRIPTION
**Pull request by AlexeyKhrabrov**
This PR makes the following changes:

- A third index is added to the class loader table - name of the first class loaded by the given class loader. This mapping will be used to find the right class loader to lookup classes in when deserializing cached AOT methods received from the JITServer. The mapping between class loaders and the name of the first loaded class is maintained only if the JVM uses the JITServer and the new runtime option `-XX:+JITServerUseAOTCache` is specified (it's disabled by default).

- The multi-index map is now more space-efficient: the same entry struct is linked into multiple lists (one for each hash table). Removing an entry when a class loader is unloaded is now O(1) as it doesn't scan the whole class chain hash table. The implementation is refactored to avoid duplicating the code for manipulating the hash tables.

- Added comments describing how the access to the class loader table is synchronized, and runtime assertions to check the necessary assumptions. Fixed a missing memory barrier.

Issue: #12153

### CODE REVIEW

**Comment by AlexeyKhrabrov** [comment]
@mpirvu Can I please get a review for this PR? Thanks.

**Comment by dsouzai** [comment]
> A third index is added to the class loader table - name of the first class loaded by the given class loader. 

Don't we already have this? When we call
```
TR_PersistentClassLoaderTable::associateClassLoaderWithClass(void *classLoaderPointer, TR_OpaqueClassBlock *clazz)
```
we associate the `classLoaderPointer` with the class chain of `clazz`; the first entry in the class chain is the offset to the `J9ROMClass` of `clazz`. So if you have the class chain of a class, you have its J9ROMClass and thus you have its name. So why we need to have a different association with the name?

**Comment by AlexeyKhrabrov** [comment]
>  So why we need to have a different association with the name?

To do the reverse lookup: given the name of the first loaded class, find the associated class loader. This will be done when deserializing an AOT method received from the JITServer that was originally compiled for a JVM with an entirely different SCC. We have to use the name of the first loaded class as a way to identify class loaders across different SCCs.

I agree that a separate loader -> name lookup function is not necessary. I will update the code to remove it - the less code the better. Same applies to the `_name` pointer in `TR_ClassLoaderInfo`: it can be retrieved on demand using the class chain, so it's not necessary to store it.

**Comment by dsouzai** [comment]
>To do the reverse lookup: given the name of the first loaded class, find the associated class loader. This will be done when deserializing an AOT method received from the JITServer that was originally compiled for a JVM with an entirely different SCC. We have to use the name of the first loaded class as a way to identify class loaders across different SCCs.

So the idea here is that we request a compilation; to do so the client gives the JITServer instance a `J9Method *`. The server figures out that there's a cached AOT method that we can send back. So it sends it back. A couple of questions about that:

1. What exactly are you looking up the classloaders for?
2. If all of the offsets in the various classchains that will be referenced in the validation records are wrt to the the Server SCC, how will the client do validation?

**Comment by AlexeyKhrabrov** [comment]
I will add a brief description of the whole mechanism in the issue #12153 (hopefully by tomorrow). For now, to answer the questions:

>     1. What exactly are you looking up the classloaders for?

As part of AOT method deserialization, the client needs to find ROMClasses in its SCC by name. In the general case, this requires a class loader instance.

>     2. If all of the offsets in the various classchains that will be referenced in the validation records are wrt to the the Server SCC, how will the client do validation?

The server doesn't have an SCC. It stores serialized AOT method bodies that are detached from the SCC. The method body compiled for the original client is stored as is, but for each SCC offset in its relocation/validation records we store information required to find the corresponding entity in the new client's SCC. For ROMClasses, this information includes its name, a secure hash of its contents (to efficiently compare ROMClasses across clients), and the name of the first class loaded by its class loader. Other SCC entities, e.g. class chains, are built on top of this.

**Comment by dsouzai** [comment]
> I will add a brief description of the whole mechanism in the issue #12153 (hopefully by tomorrow). 

Thank you, that will help me ask more useful questions, or will allow me to justify that everything is equivalent to the non-JITServer AOT semantics.

**Comment by AlexeyKhrabrov** [comment]
Updated the code to remove the unnecessary separate loader -> name lookup.

**Comment by AlexeyKhrabrov** [comment]
Updated the code to enable an assertion about synchronization assumptions under `PROD_WITH_ASSUMES` and avoid using a JITServer-specific way of getting current VM thread. This should be ready for review now, and hopefully the information in #12153 is now sufficient to reason about the changes in this PR.

**Review by mpirvu — COMMENTED** [review]
LGTM. Only a couple of nitpicks.

File: runtime/compiler/env/ClassLoaderTable.cpp [hunk 1]
```diff
@@ -20,218 +20,339 @@
...
+   // no other thread can be modifying the table at the same time.
+   TR_ASSERT(hasSharedVMAccess(vmThread), "Must have shared VM access");
+   TR_ASSERT(TR::MonitorTable::get()->getClassTableMutex()->owned_by_self(), "Must hold classTableMutex");
+   if (!_sharedCache)
+      return;
 
-   // need to acquire some kind of lock here
+   // Lookup by class loader and check if it already has an associated class
+   size_t index;
+   TR_ClassLoaderInfo *prev, *info = lookup<Loader, void *>(_loaderTable, index, prev, loader);
```
[the remark this answers is not in the document]
Comment (AlexeyKhrabrov) [reply runtime/compiler/env/ClassLoaderTable.cpp]: Done.

File: runtime/compiler/env/ClassLoaderTable.cpp [hunk 2]
```diff
@@ -20,218 +20,339 @@
...
+TR_PersistentClassLoaderTable::lookupClassLoaderAndChainAssociatedWithClassName(const uint8_t *data, size_t length) const
+   {
+   assertCurrentThreadCanRead();
+   if (!_sharedCache)
+      return { NULL, NULL };
+   size_t index;
+   TR_ClassLoaderInfo *prev, *info = lookup<Name, const uint8_t *, size_t, TR_J9SharedCache *>(
+      _nameTable, index, prev, data, length, _sharedCache
+   );
+   return { info ? info->_loader : NULL, info ? info->_chain : NULL };
```
[the remark this answers is not in the document]
Comment (AlexeyKhrabrov) [reply runtime/compiler/env/ClassLoaderTable.cpp]: Done.

**Comment by AlexeyKhrabrov** [comment]
Addressed the comments and factored out repeated calls to `J9UTF8_DATA()` and `J9UTF8_LENGTH()` to make the code more concise.

**Review by mpirvu — APPROVED** [review]

**Comment by mpirvu** [comment]
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

**Comment by mpirvu** [comment]
jenkins test sanity all jdk11

**Comment by mpirvu** [comment]
jenkins test sanity aix jdk11

**Comment by AlexeyKhrabrov** [comment]
The AIX build fails with a few errors `1540-1286 (S) The specialization does not match "template ..."` in reference to specializations of variadic templates. The version of the XLC compiler used in the build is supposed to support variadic templates (along with most other C++11 features). This might be a matter of passing a compiler flag to enable these features. It could also be a compiler bug. I'm not really able to investigate this since I don't have access to an AIX/Power setup. I could rewrite the code to avoid using variadic templates, but that would not be ideal. If JITServer is ever to be supported on AIX, this will need to be fixed. JITServer message serialization relies on heavy use of variadic templates and other C++11 features. @mpirvu your call about how to proceed.

**Comment by mpirvu** [comment]
@ymanton we would like your opinion on the issue raise by @AlexeyKhrabrov above? Does the xlc compiler on AIX support variadic templates?

**Comment by ymanton** [comment]
I can't see the options being used to invoke XLC in the build output, but based on what I see in our makefiles `-qlanglvl=extended0x` should be there, which means that whatever support XLC 13.1.3 has for C++11 in general and variadic templates specifically should already be enabled. If that's the case we might be out of luck.

I took a quick look at the code in question, it might be the specializations that are the problem rather than variadic templates in general. If you can synthesis a simple test case I can help you find an AIX host with XLC 13.1.3 installed to test on.

**Comment by AlexeyKhrabrov** [comment]
I decided to do the easier thing and changed the code to avoid using variadic templates. Ready for another review.

**Review by mpirvu — APPROVED** [review]
LGTM

**Comment by mpirvu** [comment]
jenkins test sanity all jdk11

**Comment by mpirvu** [comment]
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit jdk11

**Pull request merged**